Content.Server/Theta/ShipEvent/Systems/ShipEventFactionSystemUtils.cs

```diff
@@ -0,0 +1,252 @@
+            }
+
+            _projectileDamage[meta.EntityPrototype.ID] = damage;
+
+            return damage;
+        }
+
+        return 0;
+    }
+
+    private IPlayerSession? GetSession(EntityUid entity)
+    {
+        if (_entMan.TryGetComponent<MindComponent>(entity, out var mindComp))
+        {
+            if (mindComp.HasMind)
+            {
+                var session = mindComp.Mind!.Session;
+                if (session != null) 
+                    return session;
+            }
+        }
+
+        return null;
+    }
+
+    private IPlayerSession? GetSession(Mind.Mind mind)
+    {
+        var session = mind.Session;
+        if (session != null) 
+            return session;
+
+        return null;
+    }
+
+    public bool IsValidName(string name)
+    {
+        if (name == "") 
+            return false;
+
+        foreach (var team in Teams)
+        {
+            if (team.Name == name) 
+                return false;
+        }
+
+        return true;
+    }
+
+    public string GenerateTeamColor()
+    {
+        var failsafe = 0;
+        while (failsafe < 100)
+        {
+            failsafe++;
+            var newColor = new Color(_random.NextFloat(0, 1), _random.NextFloat(0, 1), _random.NextFloat(0, 1));
+            if (IsValidColor(newColor)) 
+                return newColor.ToHex();
+        }
+
+        return string.Empty;
+    }
+
+    public bool IsValidColor(Color color)
+    {
+        var minimalColorDelta = 200; //not based on anything, simply a magic number for comparison
+
+        foreach (var team in Teams)
+        {
+            var otherColor = Color.FromHex(team.Color);
+            var delta = RedmeanColorDelta(color, otherColor);
+            if (delta < minimalColorDelta) 
+                return false;
+        }
+
+        return true;
+    }
+
+    public bool IsValidColor(string color)
+    {
+        var newColor = Color.TryFromHex(color);
+        if (newColor == null) 
+            return false;
+
+        return IsValidColor((Color) newColor);
+    }
+
+    //todo: actually PR it to RT instead of putting it here
+    public double RedmeanColorDelta(Color a, Color b)
+    {
+        var deltaR = a.RByte - b.RByte;
+        var deltaG = a.GByte - b.GByte;
+        var deltaB = a.BByte - b.BByte;
+        var avgR = (a.RByte + b.RByte) / 2;
+        var delta = (2 + avgR / 256) * deltaR * deltaR + 4 * deltaG * deltaG + (2 + (255 - avgR) / 256) * deltaB;
+        return Math.Sqrt(delta);
+    }
+
+    public bool IsActive(EntityUid entity)
+    {
+        if (_entMan.TryGetComponent<MindComponent>(entity, out var mindComp))
+        {
+            if (mindComp.HasMind)
+            {
+                if (!mindComp.Mind!.CharacterDeadPhysically && mindComp.Mind.Session != null)
+                    return true;
+            }
+        }
+
+        return false;
+    }
+    
+    public EntityUid RandomPlace(string mapPath)
```

Comment:
```suggestion
    public EntityUid SpawnOnRandomPlace(string mapPath)
```
Название функции должно быть действием, тут же какой-то кринж

Reply:
place технически тоже глагол (i.e.: place it here)